src/main.cpp

```diff
@@ -5930,32 +5930,38 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             pfrom->fDisconnect = true;
             return true;
         }
+      
         std::vector<uint256> vtxid;
-        if (1)  // Keep this lock for as short as possible, causing 2.6 second locks
-	  {
-          LOCK2(cs_main, pfrom->cs_filter);
-          mempool.queryHashes(vtxid);
-	  }
         vector<CInv> vInv;
-        BOOST_FOREACH(uint256& hash, vtxid) {
-            CInv inv(MSG_TX, hash);
-            if (pfrom->pfilter) {
-                CTransaction tx;
-                bool fInMemPool = mempool.lookup(hash, tx);
-                if (!fInMemPool) continue; // another thread removed since queryHashes, maybe...
-                if (!pfrom->pfilter->IsRelevantAndUpdate(tx)) continue;
-            }
-            vInv.push_back(inv);
-            if (vInv.size() == MAX_INV_SZ) {
-                pfrom->PushMessage(NetMsgType::INV, vInv);
-                vInv.clear();
+        mempool.queryHashes(vtxid); // BU: No lock required here. This has it's own internal lock
+        {
+            LOCK(pfrom->cs_filter);
+            BOOST_FOREACH(uint256& hash, vtxid) {
+                CInv inv(MSG_TX, hash);
+                if (pfrom->pfilter) {
+                    CTransaction tx;
+                    // BU: the following is commented out and is not worth the effort.  If we do this will
+                    //     repeatedly lock and unlock the memory pool.
+                    //bool fInMemPool = mempool.lookup(hash, tx);
+                    //if (!fInMemPool) continue; // another thread removed since queryHashes, maybe...
+                    if (!pfrom->pfilter->IsRelevantAndUpdate(tx)) continue;
+                }
+                vInv.push_back(inv);
+                if (vInv.size() == MAX_INV_SZ) {
+                    pfrom->cs_filter.unlock();
+                    pfrom->PushMessage(NetMsgType::INV, vInv);
+                    vInv.clear();
+                    pfrom->cs_filter.lock();
```

Comment:
The unlock()...something... lock() pattern is dangerous because the lock may not be called if "something" raises an exception.  In this case PushMessage does raise exceptions.

so we basically need the inverse of LOCK... and if you look at sync.h that means we need objects like
CNoncriticalSection, CNoncriticalBlock, and then finally something like "UNLOCK()".  i don't really like the name UNLOCK() because it sounds too much like a function call (I don't like really LOCK either).  Maybe throw some indication that this is not the standard unlock behavior by calling it UNLOCKER, DELOCK, SCOPE_UNLOCKER, or something else that you prefer.